hamster_lib/objects.py

```diff
@@ -605,6 +606,78 @@ def category(self):
         """For convenience only."""
         return self.activity.category
 
+    def get_serialized_string(self):
+        """
+        Provide a canonical 'stringified' version of the fact.
+
+        This is different from ``__str__`` as we may change what information is
+        to be included in ``__str__`` anytime (and we may use localization
+        etc ..) but this property guarantees that all relevant values will be
+        encoded in the returned string in a canonical way. In that regard it
+        is in a way a counterpart to ``Fact.create_from_raw_fact``.
+        This also serves as a go-to reference implementation for 'what does a
+        complete ``raw fact`` looks like'.
+
+        Please be advised though that the ``raw_string`` used to create a
+        ``Fact`` instance is not necessarily identical to this instance's
+        ``serialized_string`` as the ``raw fact`` string may omit certain
+        values which will be autocompleted while this property always returns
+        a *complete* string.
+
+        A complete serialized fact looks like this:
+            ``2016-02-01 17:30 - 2016-02-01 18:10 making plans@world domination
+            #tag 1 #tag 2, description``
+
+            Please note that we are very liberal with allowing whitespace
+            for ``Activity.name`` and ``Category.name``.
+
+        Attention:
+            ``Fact.tags`` is a set and hence unordered. In order to provide
+            a deterministic canonical return string we will sort tags by name
+            and list them alphabetically. This is purely cosmetic and does not
+            imply any actual ordering of those facts on the instance level.
+
+        Returns:
+            text_type: Canonical string encoding all available fact info.
+        """
+        def get_times_string(fact):
+            if fact.start and not fact.end:
+                result = '{} '.format(fact.start.strftime('%Y-%m-%d %H:%M'))
+            elif fact.start and fact.end:
+                result = '{start} - {end} '.format(
+                    start=fact.start.strftime('%Y-%m-%d %H:%M'),
+                    end=fact.end.strftime('%Y-%m-%d %H:%M')
+                )
+            else:
+                result = ''
+
+            return result
+
+        category = ''
+        if self.category:
+            category = '@{}'.format(self.category.name)
+
+        tags = ''
+        if self.tags:
+            ordered_tags = sorted(list(self.tags), key=attrgetter('name'))
+            tags = ' {}'.format(' '.join(['#{}'.format(tag.name) for tag in ordered_tags]))
+
+        description = ''
+        if self.description:
+            description = ', {}'.format(self.description)
+
+        times = get_times_string(self)
+
+        result = '{times}{activity}{category}{tags}{description}'.format(
```

Comment:
I can imagine treating the activity as one variable here. Since the times have their own function, adding `get_activity_string` returning `activity[@category]` would be more consistent.